### PR TITLE
[PLT-0] Improve Overall Dev-X By Using Rye

### PR DESCRIPTION
**Pull request by adrian-chang**
Overall Goal: Improve Dev-X
Overall Strategy:
* Use `rye` as a baseline tool for all SDK Development 
   * `poetry` can be used also as standard `pyproject.toml`
* Enforces same Python / Virtual Env / Package management etc. during development versus opinionated / unstructured approach per developer.
* Refreshed CONTRIB.MD to standard CONTRIBUTING.md.
* Eliminates use of `setuptools` / `setup.py`
* Update README.MD and CONTRIB.MD to reflect use of `rye`
* Make `labelbox` package follow `src/` folder convention
* Simplify development steps (`rye run unit`, `rye run lint`, `rye run integration`) is all that is needed. Nothing else
* Make repo a monorepo by starting a `libs/` folder for future SDK modules outside of `labelbox`
  ** See comment of https://github.com/python-poetry/poetry/issues/2270, lack of monorepo support of poetry, well at least natively
* Streamline github action for deployment
* Drop tests from being in a python module environment
* Make CI integration tests use pypi built package, not local code
* Update readthedocs creation
* Test publish to pypi

TODO:
* (Eventually) Fix Readthe Docs

### CODE REVIEW

**Review by tomislav-peharda — APPROVED** [review]

File: .github/workflows/python-package.yml [hunk 1]
```diff
@@ -15,92 +19,96 @@ jobs:
...
-            da-test-key: DA_GCP_LABELBOX_API_KEY
-          - python-version: 3.9
-            prod-key: PROD_LABELBOX_API_KEY_3
-            staging-key: STAGING_LABELBOX_API_KEY_3
-            da-test-key: DA_GCP_LABELBOX_API_KEY
-          - python-version: "3.10"
-            prod-key: PROD_LABELBOX_API_KEY_4
-            staging-key: STAGING_LABELBOX_API_KEY_4
             da-test-key: DA_GCP_LABELBOX_API_KEY
+          #- python-version: 3.9
```
Comment (mnoszczak) [inline .github/workflows/python-package.yml]: Why has this been commented out?

File: libs/labelbox/README.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
+# Labelbox
+
+Core mono-module for Labelbox's SDK.
+
+## Repository Organization
+
+<<<<<<< HEAD
```
Comment (mnoszczak) [inline libs/labelbox/README.md]: ✂️

File: libs/labelbox/README.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  (e.g. Project, Label etc.). It relies on `orm/` classes for easy and succinct
+  object definitions. It also contains custom functionalities and custom GraphQL
+  templates where necessary.
+* `client.py` contains the `Client` class that's the client-side stub for
+  communicating with Labelbox servers.
+* `exceptions.py` contains declarations for all Labelbox errors.
+* `pagination.py` contains support for paginated relationship and collection
+  fetching.
+* `utils.py` contains utility functions.
+>>>>>>> c735057a (Update root readme)
```
Comment (mnoszczak) [inline libs/labelbox/README.md]: ✂️

File: libs/labelbox/README.md [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+- **`labelbox/client.py`:** Contains the `Client` class, which provides methods for interacting with the Labelbox API.
+- **`labelbox/orm/model.py`:** Defines the data model for Labelbox entities like projects, datasets, and labels.
+- **`labelbox/schema/*.py`:** Contains classes representing specific Labelbox entities and their attributes.
+- **`labelbox/data/annotation_types/*.py`:** Defines classes for different annotation types, such as bounding boxes, polygons, and classifications.
+- **`labelbox/data/serialization/*.py`:** Provides converters for different data formats.
+
+The SDK wraps the GraphQL APIs and provides a Pythonic interface for interacting with Labelbox.
+
+## Contribution guidelines
+We encourage developers to contribute to the Labelbox Python SDK and help improve its functionality and usability. Please refer to the [CONTRIB.md](CONTRIB.md) file in the root folder of the repository for detailed information on how to contribute.
```
Comment (mnoszczak) [inline libs/labelbox/README.md]: This path does not exist as the file has been moved to CONTRIBUTING.MD

File: Dockerfile [hunk 5]
```diff
@@ -1,19 +1,34 @@
...
-                gcc
-
-WORKDIR /usr/src/
-COPY requirements.txt /usr/src/
-RUN pip install -r requirements.txt
-COPY . /usr/src/
-
-RUN python setup.py install
+# https://github.com/ucyo/python-package-template/blob/master/Dockerfile
+FROM python:3.8-slim as rye
```
Comment (mnoszczak) [inline Dockerfile:2]: What's the reason behind this downgrade?

File: .github/workflows/python-package.yml (hunk 1 above)
Comment (adrian-chang) [reply .github/workflows/python-package.yml]: Accounts needed to move to internal enterprise, getting those upgraded and will put back in

File: Dockerfile (hunk 5 above)
Comment (adrian-chang) [reply Dockerfile:2]: We support down to 3.8. Testing against 3.9 (a higher version) would not be appropriate as we might assume some functionality in our tests that is only in 3.9 or higher.

**Comment by itssujan** [comment]
I havent ever used rye before.. couple of qns.
1. Is it an alternative to poetry? why not poetry?
2. On their website they say "Rye is still a very experimental tool" ..so is it stable and used widely in the industry?

File: .github/workflows/publish.yml [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        with:
+          name: build
+          path: ./dist
+      - name: Prepare package and environment
+        run: |
+          rye sync
+          rye run toml unset --toml-path pyproject.toml tool.rye.workspace
+          rye sync -f --update-all
+      - name: Integration Testing
+        env:
```
Comment (vbrodsky) [inline .github/workflows/publish.yml:102]: is there any issue if a build fails due to a flaky situation and we need to rerun the build. How would we do it? Restart the whole action?

File: .github/workflows/publish.yml [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+  
+      - name: Build and push
+        uses: docker/build-push-action@v5
+        with:
+          github-token: ${{ secrets.GITHUB_TOKEN }}
+          push: true
+  
+          platforms: |
+            linux/amd64
+            linux/arm64
```
Comment (vbrodsky) [inline .github/workflows/publish.yml:167]: this is just out of curiosity: why are we not builidng for windows (not part of this pr)

File: .github/workflows/python-package-prod.yml [hunk 8]
```diff
@@ -1,10 +1,7 @@
-name: Labelbox Python SDK Develop
+name: Labelbox Python SDK PRod (Develop)
```
Comment (vbrodsky) [inline .github/workflows/python-package-prod.yml]: what is `PRod`?

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+```
+After the above, running `git commit ...` will attempt to fix formatting,
+and make necessary changes to files. You will then need to stage those files again.
+
+You may also manually format your code by running the following:
+```bash
+rye run lint
+```
+
+### Documentation
```
Comment (vbrodsky) [inline CONTRIBUTING.md]: I wonder what are the changes to our sdk release process documented here https://labelbox.atlassian.net/wiki/spaces/PLT/pages/1721139216/Release+a+new+Python+SDK+Version#Step-4%3A-Update-readthedocs-and-verify-package-in-pypi but also elsewhere in the doc

**Review by vbrodsky — COMMENTED** [review]
Very good! Left some questions

File: .github/workflows/publish.yml [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+          rye run toml unset --toml-path pyproject.toml tool.rye.workspace
+          rye sync -f --update-all
+      - name: Integration Testing
+        env:
+          PYTEST_XDIST_AUTO_NUM_WORKERS: 32
+          LABELBOX_TEST_API_KEY: ${{ secrets[matrix.prod-key] }}
+          DA_GCP_LABELBOX_API_KEY: ${{ secrets[matrix.da-test-key] }}
+          LABELBOX_TEST_ENVIRON: prod
+        run: |
+          rye add labelbox --path ./$(find ./dist/ -name *.tar.gz) --sync
```
Comment (itssujan) [inline .github/workflows/publish.yml:108]: Just adding it as a review comment so that it doesnt get missed out


I havent ever used rye before.. couple of qns.

Is it an alternative to poetry? why not poetry?
On their website they say "Rye is still a very experimental tool" ..so is it stable and used widely in the industry?

Comment (adrian-chang) [reply .github/workflows/publish.yml:108]: https://alpopkes.com/posts/python/packaging_tools/

^ look at the comparison, I personally don't want to deal with inconsistent environments (poetry doesn't solve the python-version problem, you'd have to use pyenv, something else)

https://astral.sh/blog/uv -> for corporate sponsorship

Star History Chart](https://star-history.com/#astral-sh/rye&python-poetry/poetry&Date) -> popularity, well I mean the creator did make Flask (if that is meaningful reputation wise).

https://rye-up.com/philosophy/

You can use poetry since, by definition, the pyproject.toml files are pep standard. Not against poetry, in fact, I've used / use it in other projects outside of LB.

example (open ai): https://github.com/openai/openai-python/blob/main/requirements.lock, codeinterpreter-api (https://github.com/shroominic/codeinterpreter-api/blob/main/requirements.lock)

To your point, there is some language (last year) indicating a more experimental attitude towards the work but the recent website updates are starting to eliminate this language.

File: .github/workflows/python-package-prod.yml (hunk 8 above)
Comment (adrian-chang) [reply .github/workflows/python-package-prod.yml]: Typo sorry

File: .github/workflows/publish.yml (hunk 7 above)
Comment (adrian-chang) [reply .github/workflows/publish.yml:167]: This is currently only being used for SDK tests (codefresh)

If there's external developer need for windows, we could try and build a windows image

File: .github/workflows/publish.yml (hunk 6 above)
Comment (adrian-chang) [reply .github/workflows/publish.yml:102]: Done at the moment split up the workflow actions to rerun individual jobs

File: CONTRIBUTING.md (hunk 9 above)
Comment (adrian-chang) [reply CONTRIBUTING.md]: Publishing isn't documented in CONTRIBUTING.md (for us to publish), I don't think there's a concern though and this is something we can revisit over time
